### PR TITLE
ScrollContainer Add a new scroll_mode to only allow scrolling in code

### PR DESCRIPTION
**Pull request by DissonantVoid**
There has been this issue with the scroll container for as far as I can remember where you cannot hide the 
scroll bars without a hacky solution and you cannot disable mouse/touch scrolling even when the scroll bars are hidden.

And while #48008 solved the first issue by allowing users to hide the scroll bars, we still cannot disable scrolling even when a scroll bar is hidden, check this example project:
[scroll container.zip](https://github.com/godotengine/godot/files/10298944/scroll.container.zip)

As you can see, setting the vertical scroll bar to `SCROLL_MODE_SHOW_NEVER` doesn't prevent users from using the mouse wheel or touch on mobile to scroll anyway, this is annoying because sometimes you would want to customize scrolling by using a UI for example (like the buttons in the example project) or simply not allow scrolling at all even if items overflow..

Another practical example I ran into in the past is when I was working on a Trello like addon, where you can organize tasks and TODOs just like in Trello, all panels where parented by one scroll container and I wanted to make it so that the only way to move around is to move the mouse close to the edge of the screen as you can see:

https://user-images.githubusercontent.com/69814029/209445434-8a8d17e0-f2b9-4e81-b419-beff25dcf446.mp4

But users can just scroll with the mouse wheel anyway even though I didn't want them to, I had no control over that

So what this PR does is allow the user to completely disable scrolling even if the container has overflowing children by introducing a new scroll mode called `SCROLL_MODE_MANUAL`, when this mode is used in vertical or horizontal scroll, the bar will become hidden, overflowing children will be hidden (unlike `SCROLL_MODE_DISABLE`) and scrolling input will be ignored (unlike `SCROLL_MODE_SHOW_NEVER`). and this should allow the user to customize how players scroll, like the idea in the video for example.

### CODE REVIEW

**Comment by DissonantVoid** [comment]
why is git showing all my previous activities on the forked repo? weird...

**Comment by Zireael07** [comment]
@DissonantVoid you probably didn't make a branch when forking, just forked to master, and made a PR from there. Or maybe something went wrong with branches. I've been in a similar situation with another fork :)

**Comment by KoBeWi** [comment]
If anything, I think this should be a separate scroll mode. A property that only works with `SCROLL_MODE_SHOW_NEVER` is hardly useful...

Also if you want a scrollable canvas with manual scrolling, you can just use PanelContainer inside a Control with clipping enabled and move its position. It's effectively the same.

**Comment by DissonantVoid** [comment]
I agree that adding a new variable for this may not be the best idea, an additional scroll mode is a better idea, something like `SCROLL_MODE_SHOW_NEVER_NO_SCROLL` or simply `SCROLL_MODE_MANUAL`

I know this can be done in other ways, it's always going to be some additional work that doesn't feel as clean nor godot-like (by godot-like I mean how godot is always simple and user friendly). like the way you suggested which requires an additional node and requires the implementation of scrolling + the tool code for editor visualization. not saying it's terribly difficult but having such functionality out of the box would come in handy!

**Comment by YuriSizov** [comment]
I don't believe a new scroll mode like that makes much sense. If you only want to scroll from code, you can just as well implement your own scroll container, it's not that hard to do, and it's a perfect opportunity to use a custom container. Godot provides a way to create them for a reason, you are not expected to only ever use built-in nodes.

I think there could be a dedicated property to disable scrolling with mouse wheel / touch. That way a user could choose to do this as a granular option. What you propose here disables most of the unique functionality of the node all together. That makes the use of the node in the first place quite questionable.

We could also implement some sort of OverflowContainer that would only handle the overflow functionality, while ScrollContainer could extend that and add scrollbars and MW/touch scrolling to the mix. You'd use it to have that stripped scroll container functionality without any scrolls or scrolling input events. Though that sounds like a proposal territory at this point.

**Comment by Chaosus** [comment]
Commits needs to be squashed properly.

**Comment by YuriSizov** [comment]
Check this document if you have issues with squashing:

https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

**Comment by DissonantVoid** [comment]
Apologies for the mess, I still don't fully understand what happened there but everything I did made even a bigger mess and it snowballed from there lol. I fixed it now, sorry..

@YuriSizov I guess you're right, a custom container can do the job.
my argument is that the fact that you cannot change the scroll behavior, like how you can't disable scrolling or change the scroll input to something else is a bit annoying to me, I think it lacks customizability in that regard. as someone who often works on ui apps I've already ran into this issue a couple of times, and maybe it's just me but I think it's more reasonable to add that option with a bit of code than to have to make a custom container that is programmatically like 90% the exact same thing as a `ScrollContainer` just so we can change the scroll behavior/input, besides I don't think it's uncommon to want different kinds of scrolling behaviors in different games/apps like having a menu that scrolls vertically but can't scroll horizontally until you unlock a new skill tree (something that would be ruined by users shift scrolling) or using UI buttons to auto scroll to next item on mobile (again users can just ignore that and scroll as they please) etc.. it's the lack of control that pushed me to make this PR.

So long story short I personally think, based on my experience so far that this is a much needed feature.

One last thing, I'm not very familiar with this process, how do you usually resolve things like this? where someone wants some feature/change while someone else don't. do you run a poll of something?

**Comment by YuriSizov** [comment]
> it's more reasonable to add that option with a bit of code than to have to make a custom container that is programmatically like 90% the exact same thing as a `ScrollContainer` just so we can change the scroll behavior/input

Scroll behavior and input handling is what the majority of the code is in there. If you want to completely disable that, your remaining implementation will be small. Containers have a dedicated method for positioning children as one of the main configuration tools. You would only need to add an extra offset in there, et voila, you have scrolled content.

Now, if you want to dynamically change between a locked state where users couldn't interact but the rest stays the same and an unlocked state where users can interact normally, then I can see a point of introducing some DISABLE_OVERFLOW mode to prevent scrolling but keep the overflow of content.

> I'm not very familiar with this process, how do you usually resolve things like this? where someone wants some feature/change while someone else don't. do you run a poll of something?

You need to open a [feature proposal](https://github.com/godotengine/godot-proposals) when suggesting a new feature or an improvement of an existing feature. We can be lenient sometimes when a PR is directly created without a proposal, but as a general rule you always must make one. A proposal is a place where you'd argue in favor of your suggested feature, provide references to previous discussions, if there were any, to demonstrate relevance, and explain the technical side of things (it helps that you've already implemented it, but that's not a requirement, especially for more complex features). Then it's public discussion.

A highly upvoted proposal doesn't necessarily mean that the feature goes in. It's still up to maintainers to decide if it makes sense, or if there is a better way, or if it doesn't fit the design of the engine. But public opinion is taken into account.

**Comment by KoBeWi** [comment]
Here's all you need to make a container with scrolling behavior:
```GDScript
@tool
extends Container

@onready var panel: Panel = $Panel

@export var scroll := Vector2():
	set(v):
		scroll = v
		queue_sort()

func _init() -> void:
	sort_children.connect(do_sort)

func do_sort():
	panel.position = Vector2()
	panel.size = size
	
	for child in panel.get_children():
		if not child is Control:
			continue
		
		child.position = -scroll
```
Example scene:
[FakeScrollContainer.tscn.txt](https://github.com/godotengine/godot/files/10308116/FakeScrollContainer.tscn.txt)
image
It's not as sophisticated as ScrollContainer, but gets the job done. The Panel node is for convenience to draw stylebox and simplify the sorting behavior.

**Comment by aaronfranke** [comment]
@DissonantVoid Is the suggested solution of making a custom container type suitable for your use case? If so, we can close this PR as it's no longer needed. Otherwise this would need to be rebased.

**Comment by DissonantVoid** [comment]
@aaronfranke I could swear I'd closed this a while ago, yeah we can totally close it to reduce pr clutter. I should've done so sooner.
